prov/efa/src/efa_msg.c

```diff
@@ -170,6 +201,8 @@ static ssize_t efa_post_recv(struct efa_ep *ep, const struct fi_msg *msg, uint64
 		return 0;
 
 	err = ibv_post_recv(qp->ibv_qp, ep->recv_more_wr_head.next, &bad_wr);
+	if (OFI_UNLIKELY(err))
```

Comment:
This is different than your first version, which only early exited for ENOMEM.  Now we're never exiting.  I don't understand why we're (eventually) returning an error, but still modifying `recv_more_wr_head`.

Reply:
The first version did not early exit either, it only changed error code.

`recv_more_wr_head` is a linked list of recv buffers to be posted, which is used for the FI_MORE implementation.

`ibv_post_recv` takes a linked list as input and will return the first receive wr that encountered error (`bad_wr`).

so we have to modify `recv_more_wr_head` even if error is encountered, because some receive might have been posted successfully. But releasing the whole list is wrong, because there might be receive buffers not posted in the list.

I think the correct error handling is to release until `bad_wr`, and set `recv_more_wr_head->next` to `bad_wr->next`. Does it sound right?